src/main/java/com/rtb/projectmanagementtool/project/ProjectController.java

```diff
@@ -0,0 +1,103 @@
+/**
+ * ProjectController.java - this file implements ProjectController, which controls the logic of
+ * working with Project objects.
+ */
+package com.rtb.projectmanagementtool.project;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.EmbeddedEntity;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+
+public class ProjectController {
+  private final String PROPERTY_USER_IDS = "userIds";
+
+  DatastoreService datastore;
+
+  public ProjectController(DatastoreService datastore) {
+    this.datastore = datastore;
+  }
+
+  /**
+   * Gets all the projects in datastore that all specified users are in
+   *
+   * @param userIds the collection of userIds whose projects are to be retrieved
+   * @return HashSet containing desired projects
+   */
+  public HashSet<ProjectData> getProjects(Collection userIds) {
+
+    HashSet<ProjectData> projectContainer = new HashSet<ProjectData>();
+
+    PreparedQuery results = datastore.prepare(new Query("Project"));
+    for (Entity entity : results.asIterable()) {
+      if (entityContainsUser(entity, userIds)) {
+        ProjectData project = new ProjectData(entity);
+        projectContainer.add(project);
+      }
+    }
+
+    return projectContainer;
+  }
+
+  /**
+   * Checks if the Project entity has at least one userId contained in userIds
+   *
+   * @param entity the project entity
+   * @param userIds the collection of userIds
+   * @return true if userIds param and entity's userIds property share at least one user
+   */
+  private boolean entityContainsUser(Entity entity, Collection userIds) {
```

Comment:
I think the concept of 'Entity' we should try to contain within the Data object.  I see what you're doing here is for checking if a user exists on a project.  As an alternative, you can go ahead and build your ProjectData (based off of the Entity) then do checks off of the ProjectData instead of the entity.

Better approach would be to filter the query itself so we don't pull down all projects every time.  But I think that'll be tricky to do if we're storing a map of {role, userID}

Reply:
Yeah, makes sense. I sent in the entity so that I wouldn't have to actually create the ProjectData object before asserting that it's okay to add to the array returned by getProjects(). If it was created and entityContainsUser() returns false it would then just be garbage in the heap. What might it look like in java code to send in a query that would need to be filtered?

Reply:
Datastore has a [filter functionality](https://cloud.google.com/appengine/docs/standard/java/datastore/queries#filters) so you wouldn't have to load all projects into memory (basically letting datastore do the heavy lifting for the filtering/sorting piece).  Although I don't think EmbeddedEntities are supported for filtering.

I think I've read somewhere that repeated fields can be filtered?  If this is the case, one approach would be to store a repeated list of strings in the format `{userID: role}` (ie json key:value format) so you can maybe do something like:
```
Filter projectAdminFilter =
    new FilterPredicate("users", FilterOperator.EQUAL, "{12345: 'ADMIN'}");
```